Implemented various Gen. 8 abilities

LOuroboros opened the pull request:
## Description
This PR provides the effect of ~~the first 6~~ many abilities defined from the Gen. 8 Games.

**EDIT**: As of right now, the abilities introduced by this PR are the following:

Intrepid Sword, Dauntless Shield, Libero, Cotton Down, Ball Fetch, Propeller Tail, Stalwart, Steam Engine, Sand Spit, Perish Body, Punk Rock, Ice Scales and Power Spot.

I do not intend to work in any other ability for the time being, so this PR is good to go minus potential corrections, as far as I am concerned.

## **Discord contact info**
Lunos#4026

ultima-soul reviewed (changes requested):
Some more issues from testing:
- Player's mon gets Intrepid Sword boost from foe's ability
    - In double battles, any activation (even from your first slot mon) seems to boost your second mon
- Same issue with Dauntless Shield
- Cotton Down on second round of activation from a multi-strike move seems to softlock
    - Ability Pop-up should be modified to stay open once #1244 is merged
- No ability pop-up for Ball Fetch

Also, you can implement Stalwart as well since it's the same as Propeller Tail.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -11990,6 +11990,11 @@ static void Cmd_handleballthrow(void)
+                if (!gHasFetchedBall)
+                {
+                    gLastUsedBall = gLastUsedItem;
+                    gBattleResults.catchAttempts[gLastUsedBall]++;
```
The Master Ball doesn't need to be taken into account.

It won't fail in a wild battle, and if lost by misusing it on a trainer battle, a Pokémon with Ball Fetch does not get it back, just like in Sword and Shield :eyes:

LOuroboros replied:
Also, that check is not redundant because the one at L11932 affects gLastUsedItem, not gLastUsedBall.

I did try to make use of that originally, but I couldn't. That is the reason why I had to add gLastUsedBall :eyes:

LOuroboros commented on src/battle_main.c:
```diff
@@ -228,6 +228,8 @@ EWRAM_DATA u32 gFieldStatuses = 0;
 EWRAM_DATA u8 gBattlerAbility = 0;
 EWRAM_DATA u16 gPartnerSpriteId = 0;
+EWRAM_DATA bool8 gHasFetchedBall = FALSE;
+EWRAM_DATA bool8 gLastUsedBall = 0;
```
May I ask what is the reasoning behind, for future reference? :eyes:

LOuroboros commented:
> Some more issues from testing:
> 
>     * Player's mon gets Intrepid Sword boost from foe's ability
>       * In double battles, any activation (even from your first slot mon) seems to boost your second mon
>     * Same issue with Dauntless Shield
>     * Cotton Down on second round of activation from a multi-strike move seems to softlock
>       * Ability Pop-up should be modified to stay open once #1244 is merged
>     * No ability pop-up for Ball Fetch
> 
> Also, you can implement Stalwart as well since it's the same as Propeller Tail.

I figured I should write up a quick update.

I have addressed every single one of those *(with the help of Blurose, Ghoulslash and Churchofcage)* **[in a separate branch](https://github.com/LOuroboros/pokeemerald/commits/gen8_abilities_BE_dev)**, except for the ability pop up behavior during Cotton Down. I'm still waiting for #1244 to be merged so I can address that :eyes:

LOuroboros commented:
> 
> 
> > Some more issues from testing:
> > ```
> > * Player's mon gets Intrepid Sword boost from foe's ability
> >   * In double battles, any activation (even from your first slot mon) seems to boost your second mon
> > * Same issue with Dauntless Shield
> > * Cotton Down on second round of activation from a multi-strike move seems to softlock
> >   * Ability Pop-up should be modified to stay open once #1244 is merged
> > * No ability pop-up for Ball Fetch
> > ```
> > 
> > 
> > Also, you can implement Stalwart as well since it's the same as Propeller Tail.
> 
> I figured I should write up a quick update.
> 
> I have addressed every single one of those _(with the help of Blurose, Ghoulslash and Churchofcage)_ **[in a separate branch](https://github.com/LOuroboros/pokeemerald/commits/gen8_abilities_BE_dev)**, except for the ability pop up behavior during Cotton Down. I'm still waiting for #1244 to be merged so I can address that 👀

#1244 was finally merged, and I was corrected the AbilityPopUp usage in Cotton Down's BattleScript.

I kept working on **[that separate branch](https://github.com/LOuroboros/pokeemerald/commits/gen8_abilities_BE_dev)** and on top of the fixes to the issues raised in this PR and adding Stalwart, I implemented some more abilities. Namely:

-Steam Engine
-Sand Spit
-Perish Body
-Punk Rock
-Ice Scales
-Power Spot *(though I feel like this one needs some extra testing.)*

I'm ready to push everything to the branch that I submitted this PR from if there are no objections :eyes:

ultima-soul commented:
Yeah, you can push those changes to this PR if you want, but obviously it'll take more time for testing.

Alternatively, you could only add changes to the abilities that were originally in this PR + Stalwart and submit the other abilities as a new PR. That would speed things up for this particular PR.

LOuroboros commented:
> 
> 
> Yeah, you can push those changes to this PR if you want, but obviously it'll take more time for testing.
> 
> Alternatively, you could only add changes to the abilities that were originally in this PR + Stalwart and submit the other abilities as a new PR. That would speed things up for this particular PR.

But it would probably still take quite some time for the new PR to be approved, so eh.

ultima-soul reviewed (commented):
Does Punk Rock really need its own flag? I thought it only would've needed the Sound flag.

LOuroboros commented:
> 
> 
> Does Punk Rock really need its own flag? I thought it only would've needed the Sound flag.

As we discussed in Discord, it certainly doesn't. Punk Rock simply needs to check for the Sound flag and leave status moves out of it.

Will fix in a moment.

LOuroboros commented:
Ready for re-review :eyes:

ultima-soul commented on src/battle_util.c:
```diff
@@ -7127,6 +7233,10 @@ static u32 CalcDefenseStat(u16 move, u8 battlerAtk, u8 battlerDef, u8 moveType,
             MulModifier(&modifier, UQ_4_12(1.5));
         break;
+    case ABILITY_PUNK_ROCK:
+        if ((gBattleMoves[move].flags & FLAG_SOUND) && !IS_MOVE_STATUS(move))
```
I think the check for being a Status move is redundant because only damaging attacks will have Damage Calc commands in their move effect script.

LOuroboros replied:
That makes sense to me. I'll remove it :eyes:

LOuroboros commented:
Ready for re-re-review :eyes:

ultima-soul commented:
Could you merge in the latest Battle Engine? This PR currently doesn't build from clean.

LOuroboros commented:
> 
> 
> Could you merge in the latest Battle Engine? This PR currently doesn't build from clean.

Done :eyes:

ghoulslash commented on data/battle_scripts_1.s:
```diff
@@ -6729,6 +6768,15 @@ BattleScript_SandstreamActivates::
+BattleScript_SandSpitActivates::
+	pause 0x20
+	call BattleScript_AbilityPopUp
+	printstring STRINGID_ASANDSTORMKICKEDUP
```
This should be `STRINGID_PKMNSXWHIPPEDUPSANDSTORM`, imo

ghoulslash commented on src/battle_util.c:
```diff
@@ -6999,6 +7102,9 @@ static u32 CalcAttackStat(u16 move, u8 battlerAtk, u8 battlerDef, u8 moveType, b
             if (gBattleMons[BATTLE_PARTNER(battlerAtk)].species == SPECIES_CHERRIM && IS_MOVE_PHYSICAL(move))
                 MulModifier(&modifier, UQ_4_12(1.5));
             break;
+        case ABILITY_POWER_SPOT:
```
according to bulbapedia, this boosts power not attack, in which case it should go under `Battery` in `CalcMoveBasePowerAfterModifiers`

LOuroboros replied:
Done :eyes:

ultima-soul reviewed (changes requested):
The error with gBattleResults.catchAttempts access is still there, which I believe we've discussed how to fix on Discord.

LOuroboros commented:
Ready for... uh... I don't know, re-re-re-re-re-review or something 🤣

ghoulslash commented on src/battle_util.c:
```diff
@@ -4590,6 +4625,66 @@ u8 AbilityBattleEffects(u8 caseID, u8 battler, u16 ability, u8 special, u16 move
+             && gBattleMons[battler].statStages[STAT_SPEED] != 12
+             && (moveType == TYPE_FIRE || moveType == TYPE_WATER))
+            {
+                SET_STATCHANGER(STAT_SPEED, 12 - gBattleMons[battler].statStages[STAT_SPEED], FALSE);
```
should just be `SET_STATCHANGER(STAT_SPEED, 6, FALSE)` to raise it by 6 stages

ghoulslash commented on src/battle_util.c:
```diff
@@ -4590,6 +4625,66 @@ u8 AbilityBattleEffects(u8 caseID, u8 battler, u16 ability, u8 special, u16 move
+            {
+                SET_STATCHANGER(STAT_SPEED, 12 - gBattleMons[battler].statStages[STAT_SPEED], FALSE);
+                BattleScriptPushCursor();
+                gBattlescriptCurrInstr = BattleScript_TargetsStatWasMaxedOut;
```
Should be `BattleScript_TargetAbilityStatRaiseOnMoveEnd`

LOuroboros commented:
Ready for yet another re-filled review :eyes:
